File: codes/julia/wk03_ladder.jl
# Week 3 - Creation & Annihilation Operators (number-basis matrices)
# Physical Chemistry 2, SKKU (Prof. S. Joon Kwon)
# a|n> = sqrt(n)|n-1>, a'|n> = sqrt(n+1)|n+1>, [a,a'] = 1, H = a'a + 1/2
using LinearAlgebra, Printf

N = 30
a = diagm(1 => sqrt.(1.0:N-1))     # annihilation
ad = collect(a')                   # creation

# [a, a'] = 1 (exact except truncation corner)
comm = a * ad - ad * a
bulk = comm[1:N-1, 1:N-1] - I
@printf("[a,a'] = 1: max bulk error = %.2e, corner = %+.1f\n", maximum(abs.(bulk)), comm[N, N])

# Hamiltonian eigenvalues (n + 1/2)
H = ad * a + 0.5 * Matrix{Float64}(I, N, N)
ev = sort(eigvals(Symmetric(H)))
println("First 6 eigenvalues of H/hbar*omega: ", round.(ev[1:6]; digits = 6))

# X, P and [X,P] = i
X = (a + ad) / sqrt(2)
P = im .* (ad - a) / sqrt(2)
commXP = X * P - P * X
@printf("Im<0|[X,P]|0> = %.6f (expect 1)\n", imag(commXP[1, 1]))

# Matrix elements <n-1|a|n> = sqrt(n)
println("<n-1|a|n> for n=1..5: ", [round(a[n, n + 1]; digits = 4) for n in 1:5])

# Uncertainty product in |n>: sigma_x sigma_p = (n + 1/2) hbar
for n in (0, 1, 4)
    v = zeros(N); v[n + 1] = 1.0
    sx = sqrt(v' * (X * X) * v - (v' * X * v)^2)
    sp = sqrt(real(v' * (P * P) * v - (v' * P * v)^2))
    @printf("n=%d: sigma_x*sigma_p = %.5f hbar (expect %.1f)\n", n, real(sx * sp), n + 0.5)
end

File: codes/python/wk03_ladder.py
"""
Week 3 - Creation & Annihilation (Ladder) Operators for the SHO
Physical Chemistry 2, SKKU (Prof. S. Joon Kwon)

Builds a, a^dagger, X, P, H as matrices in the number basis |n> and verifies:
  a|n> = sqrt(n)|n-1>,  a^dag|n> = sqrt(n+1)|n+1>,  [a, a^dag] = 1,
  H = hbar*omega (a^dag a + 1/2) with eigenvalues (n + 1/2) hbar*omega,
  [X, P] = i*hbar  (hbar = m = omega = 1).
Note: in a truncated N x N basis the identities hold exactly except
in the last row/column (truncation edge) - a nice teaching point.
"""
import numpy as np

def ladder_ops(N):
    """Return (a, adag) as N x N matrices in the number basis."""
    a = np.zeros((N, N))
    for n in range(1, N):
        a[n - 1, n] = np.sqrt(n)          # <n-1| a |n> = sqrt(n)
    return a, a.T.copy()

if __name__ == "__main__":
    N = 30
    a, ad = ladder_ops(N)

    # 1) [a, a^dag] = 1 (exact except the (N-1, N-1) truncation corner)
    comm = a @ ad - ad @ a
    I = np.eye(N)
    err_bulk = np.max(np.abs((comm - I)[:N - 1, :N - 1]))
    print(f"[a, a^dag] = 1: max bulk error = {err_bulk:.2e}, corner value = {comm[-1, -1]:+.1f}")

    # 2) Number operator & Hamiltonian
    H = ad @ a + 0.5 * I
    evals = np.sort(np.linalg.eigvalsh(H))[:6]
    print("First 6 eigenvalues of H/hbar*omega:", np.round(evals, 6))

    # 3) X, P from ladder operators; check [X, P] = i (bulk)
    X = (a + ad) / np.sqrt(2.0)
    P = 1j * (ad - a) / np.sqrt(2.0)
    commXP = X @ P - P @ X
    print(f"Im<0|[X,P]|0> = {commXP[0, 0].imag:.6f} (expect 1)")
    print(f"max bulk |[X,P] - i| = {np.max(np.abs((commXP - 1j * I)[:N-1, :N-1])):.2e}")

    # 4) Matrix elements: <n-1|a|n> = sqrt(n)
    print("a matrix elements <n-1|a|n>:", [f"{a[n-1, n]:.4f}" for n in range(1, 6)],
          "vs sqrt(n):", [f"{np.sqrt(n):.4f}" for n in range(1, 6)])

    # 5) Uncertainty product in state |n>: sigma_x * sigma_p = (n + 1/2) hbar
    for n in [0, 1, 4]:
        v = np.zeros(N); v[n] = 1.0
        sx = np.sqrt(v @ (X @ X) @ v - (v @ X @ v)**2)
        sp = np.sqrt(np.real(v @ (P @ P) @ v - (v @ P @ v)**2))
        print(f"n={n}: sigma_x*sigma_p = {sx * sp:.5f} hbar (expect {n + 0.5})")
